contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,188 @@
+---
+kep-number: 19
+title: Gang Scheduling
+authors:
+  - "@k82cn"
+owning-sig: sig-scheduling, machine-learning WG
+reviewers:
+  - "@bsalamat"
+  - "@vishh"
+approvers:
+  - "@bsalamat"
+  - "@vishh"
+editor: TBD
+creation-date: 2018-07-03
+last-updated: 2018-07-25
+status: provisional
+---
+
+# Gang Scheduling
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Function Detail](#function%20detail)
+  * [API Definitation](#api%20definition)
+  * [Lifecycle Policy](#lifecycle%20policy)
+  * [kube-arbitrator](#kube-arbitrator)
+  * [Customized Controller](#customized%20controller)
+* [Feature Interaction](#feature%20interaction)
+  * [Pod RestartPolicy](#pod%20restartpolicy)
+  * [Controller Manager](#controller%20manager)
+  * [Admission Controller](#admission%20controller)
+* [References](#references)
+
+## Motivation
+
+After the discussion at [Gang-scheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to make Gang-scheduling API in core, and implement it in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator). kube-arbitrator focuses on "batch" workload in kubernetes, and will share the same [scheduling frameworks](https://github.com/kubernetes/community/pull/2281) when it's ready. This document is used to provide definition of API object and the scheduler behaviour of gang-scheduling.
+
+## Function Detail
+
+### API Definition
+
+Although workload lifecycle requirements may be different, e.g. MPI vs. Spark, the requirements to scheduler are similar. To meet the requirements to scheduler, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodSchedulingGroup defines the scheduling requirement of a pod group
+type PodSchedulingGroup struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Spec PodSchedulingGroupTemplate
+    Status PodSchedulingGroupStatus
+}
+
+// LifecyclePolicy represents the lifecycle policy of PodSchedulingGroup
+// according to Pod's phase.
+type LifeCyclePolicy struct {
+    // The action that will be taken to the PodSchedulingGroup according to
+    // Pod's phase. One of "Restart", "None".
+    // Default to None.
+	Action Action
+
+    // The phase of pod; the controller takes actions according to this
+    // pod's phase. One of "PodFailed", "Unschedulable".
+    Event Event
```

Comment:
I think we should also add a Timeout field here. When Timeout is specified, the controller waits for the Group members to satisfy the Group requirements. If the timeout expires, it takes the `Action`.

Reply:
What does `PodFailed` mean in this context? Should this field instead be a Phase that indicates the state of the gang/group?

Reply:
> we should also add a Timeout field here.

+1, and Done

> What does `PodFailed` mean in this context? Should this field instead be a Phase that indicates the state of the gang/group?

That means at least one Pod of Group are failed; `PodGroupController` need to take action accordingly. Did not introduce `Phase` for `PodGroup` as only the Pod's "controller" knows phase; for example, for a batch job, there maybe `Failed`, `Done`, `Suspend`, `Terminated` and so on `Phase`; `PodGroupController` does not know that :(.